lib/websocket.c

```diff
@@ -85,23 +86,22 @@ static ssize_t recv_callback(wslay_event_context_ptr ctx, uint8_t *buf, size_t l
 static ssize_t send_callback(wslay_event_context_ptr ctx, const uint8_t *data, size_t len, int flags, void *_conn)
 {
     h2o_websocket_conn_t *conn = _conn;
-    h2o_iovec_t buf;
+    h2o_iovec_t *buf;
 
-    /* return WOULDBLOCK if pending (TODO: queue fixed number of chunks, instead of only one) */
-    if (h2o_socket_is_writing(conn->sock)) {
+    /* return WOULDBLOCK if pending or no buffer available */
+    if (h2o_socket_is_writing(conn->sock) ||
+        conn->_write_buf.cnt == sizeof(conn->_write_buf.bufs) / sizeof(conn->_write_buf.bufs[0])) {
         wslay_event_set_error(conn->ws_ctx, WSLAY_ERR_WOULDBLOCK);
         return -1;
     }
 
-    /* copy data */
-    conn->_write_buf = h2o_mem_realloc(conn->_write_buf, len);
-    memcpy(conn->_write_buf, data, len);
-
-    /* write */
-    buf.base = conn->_write_buf;
-    buf.len = len;
-    h2o_socket_write(conn->sock, &buf, 1, on_write_complete);
+    buf = &conn->_write_buf.bufs[conn->_write_buf.cnt];
 
+    /* copy data */
+    buf->base = h2o_mem_realloc(buf->base, len);
```

Comment:
Previously, we have kept one buffer (i.e. `conn->_write_buf`) allocated until the connection is closed. Now the PR keeps up to four buffers allocated.

Considering the fact that you care about memory footprint, does it make sense to free the allocated memory in `on_write_complete` and call `h2o_mem_alloc` here?

I am not suggesting you should; it's up to you whether we should make such change.

Reply:
yes, we can do this.

i also considered using one buf and `h2o_mem_realloc` in send_callback().
but can't make sure the memory returned by `h2o_mem_realloc` was not moved from previous alloced memory. this depending on the `len` when send_callback() was called.

so use  `h2o_mem_alloc` and freed in `on_write_complete` seems more reasonable.